File: leetcode/easy/1351_countNegatives.ts
function countNegatives(grid: number[][]): number {
    const n: number = grid.length
    const m: number = grid[0].length;

    const findNegative = (row: number[]): number => {
        let left: number = 0;
        let right: number = m - 1;
        let end: number = -1;

        while (left <= right) {
            const middle = Math.floor((left + right) / 2);

            if (row[middle] < 0) {
                end = middle;
                right = middle - 1;
            } else {
                left = middle + 1;
            }
        }

        return end;
    };

    let res: number = 0;
    for (let i: number = 0; i < n; ++i) {
        const end: number = findNegative(grid[i]);
        if (end !== -1) {
            res += m - end;
        }
    }

    return res;
};

/*
4
[
[4,3,2,-1],
[3,2,1,-1],
[1,1,-1,-2],
[-1,-1,-2,-3]
]

*/

const test = () => {
    const params = [
        {
            input: {
                grid: [[4,3,2,-1],[3,2,1,-1],[1,1,-1,-2],[-1,-1,-2,-3]],
            },
            output: 8,
        },
        {
            input: {
                grid: [[3,2],[1,0]]
            },
            output: 0,
        },
    ];

    params.forEach(({input, output}) => {
        const { grid } = input;
        const result = countNegatives(grid);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();

File: leetcode/easy/1351_countNegatives.py
from typing import List
import json
import heapq
from collections import defaultdict

class Solution:
    def countNegatives(self, grid: List[List[int]]) -> int:
        N = len(grid)
        M = len(grid[0])
        top = 0
        bottom = N-1
        negativeRow = N

        while bottom <= top:
            middle = (top + bottom) // 2
            if grid[middle][0] < 0:
                negativeRow = middle
                top = middle - 1
            else:
                bottom = middle + 1

        res = (N - negativeRow) * M

        for row in range(negativeRow):
            left = 0
            right = M-1
            negativeCol = M
            while left <= right:
                middle = (left + right) // 2
                if grid[row][middle] < 0:
                    negativeCol = middle
                    right = middle - 1
                else:
                    left = middle + 1
            res += (M - negativeCol)

        return res



def test ():
    params = [
        {
            'input': [[4,3,2,-1],[3,2,1,-1],[1,1,-1,-2],[-1,-1,-2,-3]],
            'output': 8,
        },
        {
            'input': [[3,2],[1,0]],
            'output': 0,
        },
    ]
    solution = Solution()

    for param in params:
        grid = param['input']
        result = solution.countNegatives(grid)
        print(
            'SUCCESS' if json.dumps(result) == json.dumps(param['output']) else 'ERROR',
            'input', param['input'],
            'output', param['output'],
            'result', result,
            '\n',
        )


if __name__ == '__main__':
    test()
